File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/ModelJson/DataPartitionPersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.ModelJson
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.ModelJson.types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using System.Collections.Generic;
    using System.Threading.Tasks;

    /// <summary>
    /// The data partition persistence.
    /// </summary>
    class DataPartitionPersistence
    {
        private static readonly string Tag = nameof(DataPartitionPersistence);

        public static async Task<CdmDataPartitionDefinition> FromData(CdmCorpusContext ctx, Partition obj, List<CdmTraitDefinition> extensionTraitDefList, List<CdmTraitDefinition> localExtensionTraitDefList, CdmFolderDefinition documentFolder)
        {
            var partition = ctx.Corpus.MakeObject<CdmDataPartitionDefinition>(CdmObjectType.DataPartitionDef, obj.Name);

            if (!StringUtils.IsBlankByCdmStandard(obj.Description))
                partition.Description = obj.Description;
            partition.Location = ctx.Corpus.Storage.CreateRelativeCorpusPath(
                ctx.Corpus.Storage.AdapterPathToCorpusPath(obj.Location),
                documentFolder);
            partition.RefreshTime = obj.RefreshTime;
            partition.LastFileModifiedTime = obj.LastFileModifiedTime;
            partition.LastFileStatusCheckTime = obj.LastFileStatusCheckTime;


            if (StringUtils.IsBlankByCdmStandard(partition.Location))
            {
                Logger.Warning(ctx as ResolveContext, Tag, nameof(FromData), null, CdmLogCode.WarnPersistPartitionLocMissing, partition.Name);
            }

            if (obj.IsHidden == true)
            {
                var isHiddenTrait = ctx.Corpus.MakeRef<CdmTraitReference>(CdmObjectType.TraitRef, "is.hidden", true);
                partition.ExhibitsTraits.Add(isHiddenTrait);
            }

            await Utils.ProcessAnnotationsFromData(ctx, obj, partition.ExhibitsTraits);

            var csvFormatTrait = partition.ExhibitsTraits.Item("is.partition.format.CSV") as CdmTraitReference;
            if (obj.FileFormatSettings != null)
            {
                var partitionTraitExisted = csvFormatTrait != null;
                csvFormatTrait = Utils.CreateCsvTrait(obj.FileFormatSettings, ctx, csvFormatTrait);

                if (csvFormatTrait == null) {
                    Logger.Error(ctx, Tag, nameof(FromData), null, CdmLogCode.ErrPersistCsvProcessingError);
                    return null;
                }

                if (!partitionTraitExisted)
                {
                    partition.ExhibitsTraits.Add(csvFormatTrait);
                }
            }
            ExtensionHelper.ProcessExtensionFromJson(ctx, obj, partition.ExhibitsTraits, extensionTraitDefList, localExtensionTraitDefList);

            return partition;
        }

        public static async Task<Partition> ToData(CdmDataPartitionDefinition instance, ResolveOptions resOpt, CopyOptions options)
        {
            var result = new Partition
            {
                Name = instance.Name,
                Description = instance.Description,
                Location = instance.Ctx.Corpus.Storage.CorpusPathToAdapterPath(
                    instance.Ctx.Corpus.Storage.CreateAbsoluteCorpusPath(
                        instance.Location, instance.InDocument)),
                RefreshTime = instance.RefreshTime,
                FileFormatSettings = null,
                LastFileModifiedTime = instance.LastFileModifiedTime,
                LastFileStatusCheckTime = instance.LastFileStatusCheckTime
            };

            if (result.Name == null)
            {
                Logger.Warning(instance.Ctx, Tag, nameof(ToData), instance.AtCorpusPath, CdmLogCode.WarnPersistPartitionNameNull);
                result.Name = "";
            }

            if (StringUtils.IsBlankByCdmStandard(result.Location))
            {
                Logger.Warning(instance.Ctx, Tag, nameof(ToData), instance.AtCorpusPath, CdmLogCode.WarnPersistPartitionLocMissing, result.Name);
            }

            await Utils.ProcessTraitsAndAnnotationsToData(instance.Ctx, result, instance.ExhibitsTraits);

            var t2pm = new TraitToPropertyMap(instance);

            var isHiddenTrait = t2pm.FetchTraitReference("is.hidden");
            if (isHiddenTrait != null)
            {
                result.IsHidden = true;
            }

            var csvTrait = t2pm.FetchTraitReference("is.partition.format.CSV");
            if (csvTrait != null)
            {
                var csvFormatSettings = Utils.CreateCsvFormatSettings(csvTrait);

                if (csvFormatSettings != null)
                {
                    result.FileFormatSettings = csvFormatSettings;
                    result.FileFormatSettings.Type = "CsvFormatSettings";
                }
                else
                {
                    Logger.Error(instance.Ctx, Tag, nameof(ToData), instance.AtCorpusPath, CdmLogCode.ErrPersistCsvProcessingError);
                    return null;
                }
            }

            return result;
        }
    }
}

File: objectModel/TypeScript/Persistence/ModelJson/DataPartitionPersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { ModelJson } from '..';
import {
    CdmCorpusContext,
    CdmDataPartitionDefinition,
    CdmFolderDefinition,
    cdmObjectType,
    CdmTraitDefinition,
    CdmTraitReference,
    cdmLogCode,
    copyOptions,
    Logger,
    resolveOptions,
    traitToPropertyMap,
    StringUtils
} from '../../internal';
import * as timeUtils from '../../Utilities/timeUtils';
import { processExtensionFromJson } from './ExtensionHelper';
import { CsvFormatSettings, Partition, partitionBaseProperties } from './types';

export class DataPartitionPersistence {
    private static TAG: string = DataPartitionPersistence.name;

    public static async fromData(
        ctx: CdmCorpusContext,
        object: Partition,
        extensionTraitDefList: CdmTraitDefinition[],
        localExtensionTraitDefList: CdmTraitDefinition[],
        documentFolder: CdmFolderDefinition):
        Promise<CdmDataPartitionDefinition> {
        const dataPartition: CdmDataPartitionDefinition = ctx.corpus.MakeObject(cdmObjectType.dataPartitionDef, object.name);

        if (!StringUtils.isBlankByCdmStandard(object.description)) {
            dataPartition.description = object.description;
        }

        dataPartition.location = ctx.corpus.storage.createRelativeCorpusPath(
            ctx.corpus.storage.adapterPathToCorpusPath(object.location),
            documentFolder);

        dataPartition.refreshTime = object.refreshTime;

        if (object['cdm:lastFileModifiedTime'] !== undefined) {
            dataPartition.lastFileModifiedTime = new Date(object['cdm:lastFileModifiedTime']);
        }

        if (object['cdm:lastFileStatusCheckTime'] !== undefined) {
            dataPartition.lastFileStatusCheckTime = new Date(object['cdm:lastFileStatusCheckTime']);
        }

        if (StringUtils.isBlankByCdmStandard(dataPartition.location)) {
            Logger.warning(ctx, this.TAG, this.fromData.name, null, cdmLogCode.WarnPersistPartitionLocMissing, dataPartition.name);
        }

        if (object.isHidden === true) {
            const isHiddenTrait: CdmTraitReference = ctx.corpus.MakeRef(cdmObjectType.traitRef, 'is.hidden', true);
            dataPartition.exhibitsTraits.push(isHiddenTrait);
        }

        await ModelJson.utils.processAnnotationsFromData(ctx, object, dataPartition.exhibitsTraits);

        let csvFormatTrait: CdmTraitReference = dataPartition.exhibitsTraits.item('is.partition.format.CSV') as CdmTraitReference;
        if (object.fileFormatSettings !== undefined && object.fileFormatSettings.$type === 'CsvFormatSettings') {
            const partitionTraitExisted = !!csvFormatTrait;
            csvFormatTrait = ModelJson.utils.createCsvTrait(object.fileFormatSettings, ctx, csvFormatTrait);

            if (csvFormatTrait === undefined) {
                Logger.error(ctx, this.TAG, this.fromData.name, null, cdmLogCode.ErrPersistCsvProcessingError);
                return undefined;
            }

            if (!partitionTraitExisted) {
                dataPartition.exhibitsTraits.push(csvFormatTrait);
            }
        }

        processExtensionFromJson(
            ctx,
            object,
            partitionBaseProperties,
            dataPartition.exhibitsTraits,
            extensionTraitDefList,
            localExtensionTraitDefList
        );

        return dataPartition;
    }

    public static async toData(instance: CdmDataPartitionDefinition, resOpt: resolveOptions, options: copyOptions): Promise<Partition> {
        const t2pm: traitToPropertyMap = new traitToPropertyMap(instance);
        const isHiddenTrait: CdmTraitReference = t2pm.fetchTraitReference('is.hidden');

        const result: Partition = {
            name: instance.name,
            description: instance.explanation,
            annotations: undefined,
            location: instance.ctx.corpus.storage.corpusPathToAdapterPath(
                instance.ctx.corpus.storage.createAbsoluteCorpusPath(
                    instance.location, instance.inDocument)),
            refreshTime: instance.refreshTime,
            fileFormatSettings: undefined,
            'cdm:traits': undefined,
            'cdm:lastFileModifiedTime': timeUtils.getFormattedDateString(instance.lastFileModifiedTime),
            'cdm:lastFileStatusCheckTime': timeUtils.getFormattedDateString(instance.lastFileStatusCheckTime)
        };

        if (result.name === undefined) {
            Logger.warning(instance.ctx, this.TAG, this.toData.name, instance.atCorpusPath, cdmLogCode.WarnPersistPartitionNameNull);
            result.name = '';
        }

        if (StringUtils.isBlankByCdmStandard(result.location)) {
            Logger.warning(instance.ctx, this.TAG, this.toData.name, instance.atCorpusPath, cdmLogCode.WarnPersistPartitionLocMissing, result.name);
        }

        await ModelJson.utils.processTraitsAndAnnotationsToData(instance.ctx, result, instance.exhibitsTraits);

        if (isHiddenTrait) {
            result.isHidden = true;
        }

        const csvTrait: CdmTraitReference = t2pm.fetchTraitReference('is.partition.format.CSV');

        if (csvTrait) {
            const csvFormatSettings: CsvFormatSettings = ModelJson.utils.createCsvFormatSettings(csvTrait);

            if (csvFormatSettings !== undefined) {
                result.fileFormatSettings = csvFormatSettings;
                result.fileFormatSettings.$type = 'CsvFormatSettings';
            } else {
                Logger.error(instance.ctx, this.TAG, this.toData.name, instance.atCorpusPath, cdmLogCode.ErrPersistCsvProcessingError);
                return undefined;
            }
        }

        return result;
    }
}
